lib/ClangImporter/ImportEnumInfo.cpp

```diff
@@ -69,8 +69,20 @@ void EnumInfo::classifyEnum(const clang::EnumDecl *decl,
   // underlying type of the enum, because there is no way to conjure up a
   // name for the Swift type.
   if (!decl->hasNameForLinkage()) {
-    kind = EnumKind::Constants;
-    return;
+    auto unavailableOnSwift = [](clang::Decl *decl) {
+      return decl->hasAttr<clang::AvailabilityAttr>() &&
+             decl->getAttr<clang::AvailabilityAttr>()->getUnavailable() &&
+             decl->getAttr<clang::AvailabilityAttr>()->getPlatform()->isStr("swift");
+    };
+
+    // If this enum comes from a typedef, we can find a name.
+    if (!isa<clang::TypedefType>(decl->getIntegerType().getTypePtr()) ||
+        // If the typedef is available in Swift, the user will get ambiguity.
+        // It also means they may not have intended this API to be imported like this.
+        !unavailableOnSwift(cast<clang::TypedefType>(decl->getIntegerType().getTypePtr())->getDecl())) {
```

Comment:
Not sure this is an acceptable heuristic. We may have to add another attr to make sure this is actually what the user wants. CC @beccadax

Reply:
We talked this over and we think this will work.

The deal here is that:
* Swift will import both the typedef *and* the enum.
* The typedef will be unavailable from Swift.
* Name lookup sets aside unavailable-from-Swift declarations and only uses them in diagnostics when there are no available decls, so the typedef import should be harmless.

So from the user's perspective, marking the typedef as unavailable-in-Swift means that the typedef should be set aside so the enum can be imported instead.

Reply:
@beccadax I am a little confused. If the typedef is ignored how does swift know for instance that NSEnumerationOptions=Int ?

I applied this patch and modified CFAvailability.h on my local machine and updated these module dumps:

https://github.dev/plotfi/cxx-interop-diff/pull/1

Taking a look at Foundation.NSSet it seems without the typedef we get ` func objects(withOptions opts: Int,` instead of  `func objects(withOptions opts: NSEnumerationOptions, `

Reply:
> If the typedef is ignored how does swift know for instance that NSEnumerationOptions=Int ?

We set the enum's name as the typdef's name and the base type as the typedef's type. We are able to derive this information from the enum's base type (which is the typedef, now). 

> Taking a look at Foundation.NSSet it seems without the typedef we get func objects(withOptions opts: Int, instead of func objects(withOptions opts: NSEnumerationOptions,

Yeah, I think that's the behavior we want. `opts` should be an enum (where the raw value is an `Int`), not a plain `Int`. Does that make sense?

Reply:
I agree with the parameter being a raw Int, but could there be anyway that we insert a `typealias NSEnumerationOptions = Int` and use that as the parameter type? Is NSEnumerationOptions not a raw Int here?